otherlibs/unix/channels.c

```diff
@@ -0,0 +1,90 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*             Xavier Leroy, projet Gallium, INRIA Paris                  */
+/*                                                                        */
+/*   Copyright 2017 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <caml/mlvalues.h>
+#include <caml/io.h>
+#include <caml/signals.h>
+#include "unixsupport.h"
+
+#ifdef HAS_SOCKETS
+#include <sys/socket.h>
+#include "socketaddr.h"
+#endif
+
+/* Check that the given file descriptor has "stream semantics" and
+   can therefore be used as part of buffered I/O.  Things that
+   don't have "stream semantics" include block devices and
+   UDP (datagram) sockets.
+   Returns 0 if OK, a nonzero error code if error. */
+
+static int unix_check_stream_semantics(int fd)
+{
+  struct stat buf;
+
+  if (fstat(fd, &buf) == -1) return errno;
+  switch (buf.st_mode & S_IFMT) {
```

Comment:
I think this is sufficient, but someone who can read POSIXese will have to say if one of these is set when the various `S_TYPE` macros for memory queues and semaphores and the like are true. It may be that one has to check for that separately?

Reply:
Actually, thinking about this, if none of those bits are set, the `default:` case will trigger anyway.

Reply:
As far as I know, memory queues and semaphores cannot be opened as file descriptors, e.g. `sem_open` returns a `sem_t *`, and is not made available by the OCaml Unix library.

Reply:
So my comment was inapplicable two distinct ways.